File: src/Core/Seguranca/Ataques/AtaqueEntropiaIV.php
<?php

namespace Application\Core\Seguranca\Ataques;

use Application\Core\Seguranca\AlvoCriptografico;
use Application\Core\Seguranca\AtaqueInterface;
use Application\Core\Seguranca\CriptografiaAlvo;
use Application\Core\Seguranca\ResultadoAtaque;

/**
 * O AtaqueColisaoIV já prova que os IVs não colidem numa amostra prática.
 * Esse ataque vai atrás de um sintoma diferente e mais sutil: se alguém
 * trocar `random_bytes(16)` por algo previsível mas ainda "único" (tipo
 * timestamp + contador, ou um PRNG mal semeado), a colisão pode continuar
 * rara - mas o IV vira PREVISÍVEL, o que quebra a garantia de segurança
 * mesmo sem nunca colidir de fato.
 *
 * Detecta isso com 3 sinais que um IV verdadeiramente aleatório não deveria
 * ter: bytes vizinhos correlacionados, distribuição não-uniforme por byte,
 * e sequências crescentes/monótonas entre IVs consecutivos (sintoma
 * clássico de contador ou timestamp).
 */
class AtaqueEntropiaIV implements AtaqueInterface
{
    public function __construct(private int $amostras = 2000)
    {
    }

    public function nome(): string
    {
        return 'Entropia e previsibilidade do IV';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        if (!$alvo instanceof CriptografiaAlvo) {
            throw new \Application\Core\Seguranca\SkipAtaqueException('Precisa de base64url_decode do alvo.');
        }

        $ivs = [];
        for ($i = 0; $i < $this->amostras; $i++) {
            $token = $alvo->encrypt('X');
            $decodificado = $alvo->base64urlDecode(substr($token, strlen($alvo->prefixo())));
            $ivs[] = $alvo->decompor($decodificado)['iv'];
        }

        $problemas = [];

        // Sinal 1: distribuição de bytes do IV (todas as posições, todos os IVs)
        $contagem = array_fill(0, 256, 0);
        $total = 0;
        foreach ($ivs as $iv) {
            for ($i = 0; $i < strlen($iv); $i++) {
                $contagem[ord($iv[$i])]++;
                $total++;
            }
        }
        $esperado = $total / 256;
        $qui2 = 0.0;
        foreach ($contagem as $c) {
            $qui2 += (($c - $esperado) ** 2) / $esperado;
        }
        if ($qui2 > 330) {
            $problemas[] = sprintf('distribuição de bytes suspeita (qui-quadrado=%.1f, >330 é suspeito)', $qui2);
        }

        // Sinal 2: monotonicidade - conta quantos IVs consecutivos têm o
        // primeiro byte estritamente crescente (um contador/timestamp cru
        // produziria isso quase sempre; aleatório, só ~50% das vezes).
        $crescentes = 0;
        for ($i = 1; $i < count($ivs); $i++) {
            if (ord($ivs[$i][0]) > ord($ivs[$i - 1][0])) {
                $crescentes++;
            }
        }
        $proporcaoCrescente = $crescentes / (count($ivs) - 1);
        if ($proporcaoCrescente > 0.65 || $proporcaoCrescente < 0.35) {
            $problemas[] = sprintf('primeiro byte do IV parece monotônico (%.0f%% das vezes crescente; aleatório ficaria perto de 50%%)', $proporcaoCrescente * 100);
        }

        // Sinal 3: bytes duplicados dentro do MESMO IV devem ser comuns
        // (paradoxo do aniversário para 16 bytes de 0-255 já prevê bastante
        // repetição interna) - a AUSÊNCIA de qualquer repetição interna em
        // quase todos os IVs seria estranha (sugeriria geração não-uniforme,
        // tipo bytes distintos forçados).
        $semRepeticaoInterna = 0;
        foreach ($ivs as $iv) {
            if (count(array_unique(str_split($iv))) === strlen($iv)) {
                $semRepeticaoInterna++;
            }
        }
        $proporcaoSemRepeticao = $semRepeticaoInterna / count($ivs);
        // Para 16 bytes aleatórios de 0-255, a chance de TODOS distintos é ~72%.
        if ($proporcaoSemRepeticao > 0.90 || $proporcaoSemRepeticao < 0.50) {
            $problemas[] = sprintf('%.0f%% dos IVs não têm nenhum byte repetido internamente (esperado ~72%% para 16 bytes aleatórios)', $proporcaoSemRepeticao * 100);
        }

        if (!empty($problemas)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'alta',
                detalhes: implode('; ', $problemas),
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: sprintf(
                'qui-quadrado=%.1f, %.0f%% primeiro-byte-crescente (~50%% esperado), %.0f%% sem repetição interna (~72%% esperado) - tudo consistente com IV aleatório',
                $qui2,
                $proporcaoCrescente * 100,
                $proporcaoSemRepeticao * 100
            ),
        );
    }
}

File: packages/go/internal/seguranca/ataques/ataque_entropia_iv.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesEntropiaIV configura o AtaqueEntropiaIV.
type OpcoesEntropiaIV struct {
	Amostras int
}

// AtaqueEntropiaIV detecta previsibilidade no IV com três sinais: distribuição
// de bytes, monotonicidade do primeiro byte e repetição interna.
type AtaqueEntropiaIV struct {
	amostras int
}

// NovoAtaqueEntropiaIV cria o ataque. Sem argumento, usa amostras = 2000.
func NovoAtaqueEntropiaIV(opcoes ...OpcoesEntropiaIV) *AtaqueEntropiaIV {
	o := OpcoesEntropiaIV{Amostras: 2000}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueEntropiaIV{amostras: o.Amostras}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueEntropiaIV) Nome() string {
	return "Entropia e previsibilidade do IV"
}

// Executar roda o ataque.
func (a *AtaqueEntropiaIV) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	ivs := make([][]byte, 0, a.amostras)
	for i := 0; i < a.amostras; i++ {
		token, err := alvo.Encrypt([]byte("X"))
		if err != nil {
			continue
		}
		decodificado, err := alvo.Base64urlDecode(token[len(alvo.Prefixo()):])
		if err != nil {
			continue
		}
		ivs = append(ivs, alvo.Decompor(decodificado).IV)
	}

	problemas := []string{}

	// Sinal 1: distribuição de bytes do IV (todas as posições, todos os IVs)
	contagem := make([]int, 256)
	total := 0
	for _, iv := range ivs {
		for i := 0; i < len(iv); i++ {
			contagem[iv[i]]++
			total++
		}
	}
	esperado := float64(total) / 256
	qui2 := 0.0
	for _, c := range contagem {
		diff := float64(c) - esperado
		qui2 += (diff * diff) / esperado
	}
	if qui2 > 330 {
		problemas = append(problemas, fmt.Sprintf("distribuição de bytes suspeita (qui-quadrado=%.1f, >330 é suspeito)", qui2))
	}

	// Sinal 2: monotonicidade do primeiro byte
	crescentes := 0
	for i := 1; i < len(ivs); i++ {
		if ivs[i][0] > ivs[i-1][0] {
			crescentes++
		}
	}
	proporcaoCrescente := float64(crescentes) / float64(len(ivs)-1)
	if proporcaoCrescente > 0.65 || proporcaoCrescente < 0.35 {
		problemas = append(problemas, fmt.Sprintf(
			"primeiro byte do IV parece monotônico (%.0f%% das vezes crescente; aleatório ficaria perto de 50%%)",
			proporcaoCrescente*100,
		))
	}

	// Sinal 3: bytes duplicados dentro do MESMO IV
	semRepeticaoInterna := 0
	for _, iv := range ivs {
		unicos := map[byte]bool{}
		for _, b := range iv {
			unicos[b] = true
		}
		if len(unicos) == len(iv) {
			semRepeticaoInterna++
		}
	}
	proporcaoSemRepeticao := float64(semRepeticaoInterna) / float64(len(ivs))
	if proporcaoSemRepeticao > 0.9 || proporcaoSemRepeticao < 0.5 {
		problemas = append(problemas, fmt.Sprintf(
			"%.0f%% dos IVs não têm nenhum byte repetido internamente (esperado ~72%% para 16 bytes aleatórios)",
			proporcaoSemRepeticao*100,
		))
	}

	if len(problemas) > 0 {
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevAlta,
			Detalhes:   strings.Join(problemas, "; "),
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes: fmt.Sprintf(
			"qui-quadrado=%.1f, %.0f%% primeiro-byte-crescente (~50%% esperado), %.0f%% sem repetição interna (~72%% esperado) - tudo consistente com IV aleatório",
			qui2, proporcaoCrescente*100, proporcaoSemRepeticao*100,
		),
	}, nil
}
